File: typescript/core/src/transactions.ts
export type TransactionCommitment = 'processed' | 'confirmed' | 'finalized';
export type TransactionAuthScope = 'transaction:inspect' | 'transaction:send';

export interface TransactionRequestContext {
  commitment?: TransactionCommitment;
  minContextSlot?: bigint;
}

export interface LatestBlockhashResult {
  blockhash: string;
  contextSlot: bigint;
  lastValidBlockHeight: bigint;
}

export interface TransactionFeeResult {
  feeLamports: bigint | null;
  contextSlot: bigint;
}

export interface TransactionSimulationOptions extends TransactionRequestContext {
  accounts?: readonly string[];
  innerInstructions?: boolean;
  replaceRecentBlockhash?: boolean;
}

export interface TransactionSimulationResult {
  contextSlot: bigint;
  err: unknown | null;
  logs: readonly string[] | null;
  unitsConsumed?: bigint;
  /** Loaded account data size, when the relay reports it. */
  loadedAccountsDataSize?: bigint;
  accounts?: readonly unknown[] | null;
}

export interface TransactionSendOptions {
  skipPreflight?: boolean;
  preflightCommitment?: TransactionCommitment;
  minContextSlot?: bigint;
}

export interface TransactionSendResult {
  signature: string;
}

export interface TransactionSignatureStatus {
  signature: string;
  slot: bigint | null;
  confirmationStatus: TransactionCommitment | null;
  err: unknown | null;
}

export interface TransactionTransportErrorBody {
  code: string;
  message: string;
  retryable: boolean;
  request_id?: string;
  submission_state?: 'not_submitted' | 'unknown';
  signature?: string;
  details?: unknown;
}

export class TransactionTransportError extends Error {
  readonly code: string;
  readonly retryable: boolean;
  readonly requestId?: string;
  readonly submissionState?: 'not_submitted' | 'unknown';
  readonly signature?: string;
  readonly details?: unknown;
  readonly status: number;

  constructor(status: number, body: TransactionTransportErrorBody) {
    super(body.message);
    this.name = 'TransactionTransportError';
    this.status = status;
    this.code = body.code;
    this.retryable = body.retryable;
    this.requestId = body.request_id;
    this.submissionState = body.submission_state;
    this.signature = body.signature;
    this.details = body.details;
  }
}

export interface TransactionTransport {
  getLatestBlockhash(options?: TransactionRequestContext): Promise<LatestBlockhashResult>;
  getFeeForMessage(message: string, options?: TransactionRequestContext): Promise<TransactionFeeResult>;
  simulateTransaction(
    transaction: string,
    options?: TransactionSimulationOptions
  ): Promise<TransactionSimulationResult>;
  sendTransaction(transaction: string, options?: TransactionSendOptions): Promise<TransactionSendResult>;
  getSignatureStatus(
    signature: string,
    options?: TransactionRequestContext & { searchTransactionHistory?: boolean }
  ): Promise<TransactionSignatureStatus | null>;
  getBlockHeight(options?: TransactionRequestContext): Promise<bigint>;
}

type AuthenticatedTransactionFetch = (
  input: string,
  init: RequestInit,
  scope: TransactionAuthScope,
  allowAuthReplay: boolean
) => Promise<Response>;

function decimal(value: bigint | undefined): string | undefined {
  return value === undefined ? undefined : value.toString(10);
}

function bigintField(value: unknown, field: string): bigint {
  if (typeof value !== 'string' || !/^\d+$/.test(value)) {
    throw new Error(`Invalid decimal u64 field '${field}' in transaction response`);
  }
  return BigInt(value);
}

function optionalBigint(value: unknown, field: string): bigint | undefined {
  return value === undefined || value === null ? undefined : bigintField(value, field);
}

function requestBody(value: Record<string, unknown>): string {
  return JSON.stringify(Object.fromEntries(
    Object.entries(value).filter(([, entry]) => entry !== undefined)
  ));
}

async function parseError(response: Response): Promise<TransactionTransportError> {
  let body: Partial<TransactionTransportErrorBody> = {};
  try {
    body = await response.json() as Partial<TransactionTransportErrorBody>;
  } catch {
    // Public errors are deliberately synthesized without reflecting raw bodies.
  }
  return new TransactionTransportError(response.status, {
    code: typeof body.code === 'string' ? body.code : 'transaction_transport_error',
    message: typeof body.message === 'string' ? body.message : `Transaction request failed (${response.status})`,
    retryable: body.retryable === true,
    request_id: body.request_id ?? (body as { requestId?: string }).requestId,
    submission_state: body.submission_state
      ?? (body as { submissionState?: 'not_submitted' | 'unknown' }).submissionState,
    signature: body.signature,
    details: body.details,
  });
}

export function createTransactionTransport(
  baseUrl: string,
  authenticatedFetch: AuthenticatedTransactionFetch
): TransactionTransport {
  const root = `${baseUrl.replace(/\/$/, '')}/transactions/v1`;
  const post = async <T>(
    route: string,
    body: Record<string, unknown>,
    scope: TransactionAuthScope
  ): Promise<T> => {
    const response = await authenticatedFetch(`${root}/${route}`, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: requestBody(body),
    }, scope, true);
    if (!response.ok) throw await parseError(response);
    return response.json() as Promise<T>;
  };

  return {
    async getLatestBlockhash(options = {}) {
      const value = await post<Record<string, unknown>>('latest-blockhash', {
        commitment: options.commitment,
        minContextSlot: decimal(options.minContextSlot),
      }, 'transaction:inspect');
      return {
        blockhash: String(value.blockhash),
        contextSlot: bigintField(value.contextSlot, 'contextSlot'),
        lastValidBlockHeight: bigintField(value.lastValidBlockHeight, 'lastValidBlockHeight'),
      };
    },
    async getFeeForMessage(message, options = {}) {
      const value = await post<Record<string, unknown>>('fee', {
        message,
        commitment: options.commitment,
        minContextSlot: decimal(options.minContextSlot),
      }, 'transaction:inspect');
      return {
        feeLamports: value.feeLamports === null ? null : bigintField(value.feeLamports, 'feeLamports'),
        contextSlot: bigintField(value.contextSlot, 'contextSlot'),
      };
    },
    async simulateTransaction(transaction, options = {}) {
      const value = await post<Record<string, unknown>>('simulate', {
        transaction,
        commitment: options.commitment,
        minContextSlot: decimal(options.minContextSlot),
        accounts: options.accounts ? { addresses: options.accounts } : undefined,
        innerInstructions: options.innerInstructions,
        replaceRecentBlockhash: options.replaceRecentBlockhash,
      }, 'transaction:inspect');
      return {
        contextSlot: bigintField(value.contextSlot, 'contextSlot'),
        err: value.err ?? null,
        logs: value.logs as readonly string[] | null ?? null,
        unitsConsumed: optionalBigint(value.unitsConsumed, 'unitsConsumed'),
        loadedAccountsDataSize: optionalBigint(
          value.loadedAccountsDataSize,
          'loadedAccountsDataSize'
        ),
        accounts: value.accounts as readonly unknown[] | null | undefined,
      };
    },
    sendTransaction(transaction, options = {}) {
      return post<TransactionSendResult>('send', {
        transaction,
        skipPreflight: options.skipPreflight,
        preflightCommitment: options.preflightCommitment,
        minContextSlot: decimal(options.minContextSlot),
      }, 'transaction:send');
    },
    async getSignatureStatus(signature, options = {}) {
      const value = await post<Record<string, unknown> | null>('signature-status', {
        signature,
        searchTransactionHistory: options.searchTransactionHistory,
      }, 'transaction:inspect');
      const status = value?.status as Record<string, unknown> | null | undefined;
      if (!status) return null;
      return {
        signature,
        slot: status.slot === null ? null : bigintField(status.slot, 'slot'),
        confirmationStatus: status.confirmationStatus as TransactionCommitment | null,
        err: status.err ?? null,
      };
    },
    async getBlockHeight(options = {}) {
      const value = await post<Record<string, unknown>>('block-height', {
        commitment: options.commitment,
        minContextSlot: decimal(options.minContextSlot),
      }, 'transaction:inspect');
      return bigintField(value.blockHeight, 'blockHeight');
    },
  };
}

File: python/arete-sdk/arete/transactions.py
"""Transaction relay transport (``POST <base>/transactions/v1/*``).

Port of ``typescript/core/src/transactions.ts`` / Rust
``arete_sdk::transactions``: the :class:`TransactionTransport` protocol over
the six relay routes and :class:`HttpTransactionTransport`, authenticated via
:mod:`arete.http` with ``transaction:inspect`` / ``transaction:send`` scopes.
``send`` requests carry the predispatch marker and only replay after a token
refresh when the server proved the upstream dispatch was never attempted.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Protocol, Sequence

from arete.errors import AreteError
from arete.http import AuthTokenTarget, HttpAuthClient, HttpRequestError

INSPECT_SCOPE = "transaction:inspect"
SEND_SCOPE = "transaction:send"

_COMMITMENTS = ("processed", "confirmed", "finalized")


class TransactionTransportError(AreteError):
    """Structured transaction relay failure (full TS error body)."""

    def __init__(
        self,
        status: int,
        *,
        code: str,
        message: str,
        retryable: bool = False,
        request_id: Optional[str] = None,
        submission_state: Optional[str] = None,  # 'not_submitted' | 'unknown'
        signature: Optional[str] = None,
        details: Any = None,
    ) -> None:
        super().__init__(message)
        self.status = status
        self.code = code
        self.message = message
        self.retryable = retryable
        self.request_id = request_id
        self.submission_state = submission_state
        self.signature = signature
        self.details = details


@dataclass(frozen=True)
class LatestBlockhashResult:
    blockhash: str
    context_slot: int
    last_valid_block_height: int


@dataclass(frozen=True)
class TransactionFeeResult:
    fee_lamports: Optional[int]
    context_slot: int


@dataclass(frozen=True)
class TransactionSimulationResult:
    context_slot: int
    err: Any = None
    logs: Optional[List[str]] = None
    units_consumed: Optional[int] = None
    accounts: Optional[List[Any]] = None
    # Loaded account data size the simulated transaction touched, in bytes.
    # Absent/null stays ``None``; ``"0"`` parses to ``0`` (the distinction is
    # load-bearing for transaction V1 budget estimation, SIMD-0385).
    loaded_accounts_data_size: Optional[int] = None


@dataclass(frozen=True)
class TransactionSendResult:
    signature: str


@dataclass(frozen=True)
class TransactionSignatureStatus:
    signature: str
    slot: Optional[int]
    confirmation_status: Optional[str]
    err: Any = None


class TransactionTransport(Protocol):
    async def get_latest_blockhash(
        self,
        *,
        commitment: Optional[str] = None,
        min_context_slot: Optional[int] = None,
    ) -> LatestBlockhashResult: ...

    async def get_fee_for_message(
        self,
        message: str,
        *,
        commitment: Optional[str] = None,
        min_context_slot: Optional[int] = None,
    ) -> TransactionFeeResult: ...

    async def simulate_transaction(
        self,
        transaction: str,
        *,
        commitment: Optional[str] = None,
        min_context_slot: Optional[int] = None,
        accounts: Optional[Sequence[str]] = None,
        inner_instructions: Optional[bool] = None,
        replace_recent_blockhash: Optional[bool] = None,
    ) -> TransactionSimulationResult: ...

    async def send_transaction(
        self,
        transaction: str,
        *,
        skip_preflight: Optional[bool] = None,
        preflight_commitment: Optional[str] = None,
        min_context_slot: Optional[int] = None,
    ) -> TransactionSendResult: ...

    async def get_signature_status(
        self,
        signature: str,
        *,
        search_transaction_history: Optional[bool] = None,
    ) -> Optional[TransactionSignatureStatus]: ...

    async def get_block_height(
        self,
        *,
        commitment: Optional[str] = None,
        min_context_slot: Optional[int] = None,
    ) -> int: ...


def _decimal(value: Optional[int]) -> Optional[str]:
    if value is None:
        return None
    if not isinstance(value, int) or isinstance(value, bool) or value < 0:
        raise ValueError("minContextSlot must be a non-negative integer")
    return str(value)


def _int_field(value: Any, name: str) -> int:
    if not isinstance(value, str) or not value or not value.isascii() or not value.isdigit():
        raise TransactionTransportError(
            0,
            code="invalid_response",
            message=f"Invalid decimal u64 field '{name}' in transaction response",
        )
    return int(value)


def _optional_int(value: Any, name: str) -> Optional[int]:
    if value is None:
        return None
    return _int_field(value, name)


def _request_body(entries: Dict[str, Any]) -> Dict[str, Any]:
    return {key: value for key, value in entries.items() if value is not None}


def _transport_error(status: Optional[int], body: Any) -> TransactionTransportError:
    parsed = body if isinstance(body, dict) else {}
    status = status or 0
    # Public errors are deliberately synthesized without reflecting raw bodies.
    return TransactionTransportError(
        status,
        code=parsed.get("code")
        if isinstance(parsed.get("code"), str)
        else "transaction_transport_error",
        message=parsed.get("message")
        if isinstance(parsed.get("message"), str)
        else f"Transaction request failed ({status})",
        retryable=parsed.get("retryable") is True,
        request_id=parsed.get("request_id", parsed.get("requestId")),
        submission_state=parsed.get("submission_state", parsed.get("submissionState")),
        signature=parsed.get("signature"),
        details=parsed.get("details"),
    )


class HttpTransactionTransport:
    """HTTP :class:`TransactionTransport` over ``<base>/transactions/v1``."""

    def __init__(
        self,
        base_url: str,
        auth_client: HttpAuthClient,
        *,
        target: Optional[AuthTokenTarget] = None,
    ) -> None:
        self._root = base_url.rstrip("/") + "/transactions/v1"
        self._auth = auth_client
        self._target = target

    async def _post(
        self, route: str, body: Dict[str, Any], scope: str
    ) -> Any:
        try:
            return await self._auth.request_json(
                "POST",
                f"{self._root}/{route}",
                json_body=_request_body(body),
                target=self._target,
                scopes=(scope,),
            )
        except HttpRequestError as e:
            raise _transport_error(
                getattr(e, "status", None), getattr(e, "body", None)
            ) from e

    async def get_latest_blockhash(
        self,
        *,
        commitment: Optional[str] = None,
        min_context_slot: Optional[int] = None,
    ) -> LatestBlockhashResult:
        value = await self._post(
            "latest-blockhash",
            {"commitment": commitment, "minContextSlot": _decimal(min_context_slot)},
            INSPECT_SCOPE,
        )
        return LatestBlockhashResult(
            blockhash=str(value["blockhash"]),
            context_slot=_int_field(value.get("contextSlot"), "contextSlot"),
            last_valid_block_height=_int_field(
                value.get("lastValidBlockHeight"), "lastValidBlockHeight"
            ),
        )

    async def get_fee_for_message(
        self,
        message: str,
        *,
        commitment: Optional[str] = None,
        min_context_slot: Optional[int] = None,
    ) -> TransactionFeeResult:
        value = await self._post(
            "fee",
            {
                "message": message,
                "commitment": commitment,
                "minContextSlot": _decimal(min_context_slot),
            },
            INSPECT_SCOPE,
        )
        fee = value.get("feeLamports")
        return TransactionFeeResult(
            fee_lamports=None if fee is None else _int_field(fee, "feeLamports"),
            context_slot=_int_field(value.get("contextSlot"), "contextSlot"),
        )

    async def simulate_transaction(
        self,
        transaction: str,
        *,
        commitment: Optional[str] = None,
        min_context_slot: Optional[int] = None,
        accounts: Optional[Sequence[str]] = None,
        inner_instructions: Optional[bool] = None,
        replace_recent_blockhash: Optional[bool] = None,
    ) -> TransactionSimulationResult:
        value = await self._post(
            "simulate",
            {
                "transaction": transaction,
                "commitment": commitment,
                "minContextSlot": _decimal(min_context_slot),
                "accounts": {"addresses": list(accounts)} if accounts else None,
                "innerInstructions": inner_instructions,
                "replaceRecentBlockhash": replace_recent_blockhash,
            },
            INSPECT_SCOPE,
        )
        return TransactionSimulationResult(
            context_slot=_int_field(value.get("contextSlot"), "contextSlot"),
            err=value.get("err"),
            logs=value.get("logs"),
            units_consumed=_optional_int(value.get("unitsConsumed"), "unitsConsumed"),
            accounts=value.get("accounts"),
            loaded_accounts_data_size=_optional_int(
                value.get("loadedAccountsDataSize"), "loadedAccountsDataSize"
            ),
        )

    async def send_transaction(
        self,
        transaction: str,
        *,
        skip_preflight: Optional[bool] = None,
        preflight_commitment: Optional[str] = None,
        min_context_slot: Optional[int] = None,
    ) -> TransactionSendResult:
        value = await self._post(
            "send",
            {
                "transaction": transaction,
                "skipPreflight": skip_preflight,
                "preflightCommitment": preflight_commitment,
                "minContextSlot": _decimal(min_context_slot),
            },
            SEND_SCOPE,
        )
        return TransactionSendResult(signature=str(value["signature"]))

    async def get_signature_status(
        self,
        signature: str,
        *,
        search_transaction_history: Optional[bool] = None,
    ) -> Optional[TransactionSignatureStatus]:
        value = await self._post(
            "signature-status",
            {
                "signature": signature,
                "searchTransactionHistory": search_transaction_history,
            },
            INSPECT_SCOPE,
        )
        status = value.get("status") if isinstance(value, dict) else None
        if not status:
            return None
        slot = status.get("slot")
        return TransactionSignatureStatus(
            signature=signature,
            slot=None if slot is None else _int_field(slot, "slot"),
            confirmation_status=status.get("confirmationStatus"),
            err=status.get("err"),
        )

    async def get_block_height(
        self,
        *,
        commitment: Optional[str] = None,
        min_context_slot: Optional[int] = None,
    ) -> int:
        value = await self._post(
            "block-height",
            {"commitment": commitment, "minContextSlot": _decimal(min_context_slot)},
            INSPECT_SCOPE,
        )
        return _int_field(value.get("blockHeight"), "blockHeight")
